airflow/providers/apache/beam/operators/beam.py

```diff
@@ -470,3 +529,144 @@ def on_kill(self) -> None:
+    ]
+    template_fields_renderers = {'dataflow_config': 'json', 'pipeline_options': 'json'}
+
+    def __init__(
+        self,
+        *,
+        go_file: str,
+        runner: str = "DirectRunner",
+        default_pipeline_options: Optional[dict] = None,
+        pipeline_options: Optional[dict] = None,
+        gcp_conn_id: str = "google_cloud_default",
+        delegate_to: Optional[str] = None,
+        dataflow_config: Optional[Union[DataflowConfiguration, dict]] = None,
+        **kwargs,
+    ) -> None:
+        super().__init__(
+            runner=runner,
+            default_pipeline_options=default_pipeline_options,
+            pipeline_options=pipeline_options,
+            gcp_conn_id=gcp_conn_id,
+            delegate_to=delegate_to,
+            dataflow_config=dataflow_config,
+            **kwargs,
+        )
+
+        self.go_file = go_file
+        self.should_init_go_module = False
+
+    def execute(self, context):
+        """Execute the Apache Beam Pipeline."""
+        (
+            is_dataflow,
+            dataflow_job_name,
+            formatted_pipeline_options,
+            process_line_callback,
+        ) = self._init_pipeline_options(format_pipeline_options=True, job_name_variable_key="job_name")
+
+        with ExitStack() as exit_stack:
+            if self.go_file.lower().startswith("gs://"):
+                gcs_hook = GCSHook(self.gcp_conn_id, self.delegate_to)
+                # Go doesn't allow `go mod tidy` inside root tmp dir, creating a go subdir to put files inside
+                if not os.path.isdir("/tmp/go"):
+                    os.mkdir("/tmp/go")
+                tmp_gcs_file = exit_stack.enter_context(
+                    gcs_hook.provide_file(object_url=self.go_file, dir="/tmp/go")
+                )
+                self.go_file = tmp_gcs_file.name
+                self.should_init_go_module = True
+
+            if is_dataflow:
+                with self.dataflow_hook.provide_authorized_gcloud():
+                    self.beam_hook.start_go_pipeline(
+                        variables=formatted_pipeline_options,
+                        go_file=self.go_file,
+                        process_line_callback=process_line_callback,
+                        should_init_module=self.should_init_go_module,
+                    )
+
+                self.dataflow_hook.wait_for_done(
+                    job_name=dataflow_job_name,
+                    location=self.dataflow_config.location,
+                    job_id=self.dataflow_job_id,
+                    multiple_jobs=False,
+                )
+
+            else:
+                self.beam_hook.start_go_pipeline(
+                    variables=formatted_pipeline_options,
+                    go_file=self.go_file,
+                    process_line_callback=process_line_callback,
+                    should_init_module=self.should_init_go_module,
+                )
+
+        return {"dataflow_job_id": self.dataflow_job_id}
```

Comment:
should this be inside is dataflow case? what is dataflow job id otherwise?

Reply:
If the runner is not `dataflow`, this will default to `None`. Indeed I don't make much sense to return a `dataflow_job_id` when we are not in the dataflow case. I wonder if this is for backward compatibility reasons.